framework/include/geomsearch/NearestNodeLocator.h

```diff
@@ -94,6 +100,7 @@ class NearestNodeLocator : public Restartable
   BoundaryID _boundary2;
 
   bool _first;
+  bool _ghost_elements;
```

Comment:
I added the _ghost_elements variable so that the ghosting happens only once during the entire simulation. Without this variable, the entire master and slave boundary would be ghosted every time the patch is updated using the `patch_update_strategy`. 

Yes, I will conduct timing comparisons on a fuel problem.

Reply:
I tried a bison example problem with 320 elements in the slave and master surface with auto patch update strategy with patch size of 5 and 10 with/without ghosting the entire master and slave surfaces. 

patch size 5 - 606 s (with) / 590 s (without ghosting entire surface)
patch size 10 - 614 s (with) / 588 s (without ghosting entire surface)

So its around 2-4 % increase in time due to ghosting the entire surface.

Reply:
Oh, I see -- I missed that.

Reply:
We might see a much more significant run-time increase with ghosting the entire boundary on a larger model, so I'm a little hesitant to go with your code as-is because it changes things so that we always ghost the entire surface, whether or not we are using the new patch update strategy. If we're using one of the existing strategies, it seems like we should still just ghost the faces in the patch.

Do you think it would make sense to still use a patch just for ghosting purposes with your new strategy?  We could make it pretty big, and wouldn't have the performance penalties associated with allocating Jacobian entries, because we'd only use it for ghosting.

Reply:
I can change the code such that for the existing strategies I am ghosting only the faces in the patch. 

For the new strategy though, are you suggesting have two different patch sizes - a bigger patch size used only for ghosting the elements at the first time step, and a smaller patch size for conducting contact detection searches and adding off diag elements to the Jacobian? How would the user pick this patch size for ghosting? It would depend on the problem and a small patch size may lead to a seg fault during the nonlinear iteration patch updates due to some element face not being ghosted.

Reply:
I am going try a sliding block problem where a block of 1m length slides on another block of 30 m length. The slave and master boundaries would have 2000 and 30,000 elements, respectively. This should give a better idea of how much time ghosting the entire surface adds.

Reply:
Yeah, that's what I'm thinking -- a bigger patch for ghosting.  You don't really even need the original patch anymore, right?  If so, you could just have one patch, and that one would be for ghosting. You could check to make sure the nodes are in the ghosted patch to avoid a segfault, and instead give an error saying that the patch needs to be bigger.

Reply:
Well the original patch is technically not needed anymore. But if we don't save a patch at all, it could get expensive for problems with a large slave node set because we may end up updating the closest neighbor for all slave nodes at each nonlinear iteration.

Reply:
Yeah, we don't want to update the patch used for ghosting at every nonlinear iteration -- it doesn't even do us any good since the ghosting can only change once per timestep.

Reply:
I tried the large sliding block problem and ghosting the entire surface does take quite a bit of time in that analysis. I ran the problem with 10 processors for 5 time steps and the case without ghosting entire surface takes 69.32 s as opposed to 82.024 s for the case with ghosted surfaces. So that almost 18% increase in time.  

I am using auto patch update strategy with patch size of 5 for both the analysis.

Reply:
That doesn't surprise me.  We may even see worse effects on 3D models.

Reply:
I have changed the never, auto and always patch update strategies to use the earlier ghosting technique and the nonlinear_iter strategy ghosts the entire master and slave boundary.